Add check if we can disregard child as in set_append_rel_size

seryozhasmirnov opened the pull request:
In case of update query `subquery_planner` calls `inheritance_planner` ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planner.c#L879)) that calls `grouping_planner` ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planner.c#L1393)), in case of select query `grouping_planner` is called directly ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planner.c#L882)).

Because for update in [inheritance_planner](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planner.c#L1113-L1621) plan is generated for each child table ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planner.c#L1233-L1541) and [here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planner.c#L1392-L1393)) and [set_append_rel_size](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/path/allpaths.c#L841-L1110) function also has a loop for each child table ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/path/allpaths.c#L887-L1078)) the number of times line ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/path/allpaths.c#L955)) to call `set_dummy_rel_pathlist(root, childrel)`  in  [set_append_rel_size](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/path/allpaths.c#L841-L1110) is called (NumberOfChildTables)*(NumberOfChildTables-1) times.

That leads to the high memory consumption if there are too many partitions.

This fix uses logic from [set_append_rel_size](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/path/allpaths.c#L841-L1110) to check if the child need to be scanned ([here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/path/allpaths.c#L913-L957)).
As `RelOptInfo` arrays (`root->simple_rel_array_size` and `root->simple_rel_array`) are filled in [query_planner](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planmain.c#L39-L277) that is too late, it was done in advance for `PlannerInfo` copy using the set of functions from [query_planner](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planmain.c#L39-L277) (from [here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planmain.c#L135-L153) and [here](https://github.com/arenadata/gpdb/blob/1d8968729000b55ecedf1e307f952463bc3254d6/src/backend/optimizer/plan/planmain.c#L171-L191)).

Commit:
Add check if we can disregard child as in set_append_rel_size

Stolb27 committed:
fast port changes from query_planner to inheritance_planner

Stolb27 committed:
Merge branch 'ADBDEV-2578-3' into ADBDEV-2578-2

Stolb27 committed:
fixup! merge garbage

Commit:
Add comments

andr-sokolov commented on src/backend/optimizer/plan/planner.c:
```diff
@@ -1233,11 +1236,31 @@ inheritance_planner(PlannerInfo *root)
+	 * arrays to check later if the child need to be scanned.
+	 */
+	fakeroot = palloc(sizeof(PlannerInfo));
+	memcpy(fakeroot, root, sizeof(PlannerInfo));
```
I suggest replacing memcpy with "*fakeroot = *root;". The compiler will check types of variables and the type mismatch error will be detected automatically. The error of incorrectly specifying the size of the structure will not be possible. The compiler will be able to generate more optimal code

Stolb27 committed:
fix lock upgrade issue for part tables with index

planner chooses lock level for relation indexes depending on kind
of operation on table: Row Exclusive for DML and Access Share
for SELECT. The kind of operation is detected by
root->parse->resultRelation that contains only a single rel index.
inheritance planner mutate query tree for each partition before
building a plan for it. We should handle this issue in the same
manner to avoid lock upgrading that may lead to deadlocks.

Stolb27 committed:
fixup! avoid mutate query tree twice

combine modifactions as a single code block

Stolb27 commented:
I've suggested less invasive solution in #345. Let's discuss it first.

Stolb27 commented:
Closed in favor #345
